### DISCUSSION TITLE
[Feature]: Improve error typing and management for entire react-native-video package

### DISCUSSION THREAD

**Discussion by ChristopherGabba** [TheWidlarzGroup/react-native-video#4967] [Ideas]
### What problem are you trying to solve?

Error handling in this package is very difficult in TypeScript. Currently the type for the `onError` function is difficult to work with in TypeScript. This may be related to `Readonly` but I'm not sure.

<img width="450" alt="Screenshot 2024-02-29 at 5 22 42 PM" src="https://github.com/react-native-video/react-native-video/assets/122328438/3b8c7bd0-74e9-46cd-b187-380f343a2197">

It would be awesome for intellisense to propose all the different error types available (like the screenshot from react-native-vision-camera below). This will allow the developer to gracefully handle, for the user, all the specific errors that the Video player throws.

The current error type is hard to reason about — which of these object lines contains the important data? The developer has to do a bunch of trial and error:

```javascript
type RNVideoError = { // which one of these object lines contains the important data? developer has to do a bunch of trial and error
  error: {
    errorString?: string | undefined
    errorException?: string | undefined
    errorStackTrace?: string | undefined
    errorCode?: string | undefined
    error?: string | undefined
    code?: string | undefined
    domain?: string | undefined
    localizedDescription?: string | undefined
    localizedFailureReason?: string | undefined
    localizedRecoverSuggestion?: string | undefined
  }
  target?: number | undefined
}
```

### Proposed solution

Improve error typing and management for the entire react-native-video package. Would love for all of the following:

**1. Work like this:**
```javascript
onError((error: VideoError) => /* handle error  */)
```

**2.** This error type should be importable into my code so that I don't have to manage a separate type and can use it directly.
```javascript
import { VideoError } from 'react-native-video'
```

**3. Simplify the error system to follow more typical error types.**
Currently I had to make my own type called:
```javascript
type VideoError = {
  error: {
    errorString?: string | undefined
    errorException?: string | undefined
    errorStackTrace?: string | undefined
    errorCode?: string | undefined
    error?: string | undefined
    code?: string | undefined
    domain?: string | undefined
    localizedDescription?: string | undefined
    localizedFailureReason?: string | undefined
    localizedRecoverSuggestion?: string | undefined
  }
  target?: number | undefined
}
```
But the fact is that just about every other package I use, errors are way simpler (not 12 different objects).
For example (what it should be):
```javascript
class VideoError extends Error {
   name: AllCustomErrorTypes // for example: "EmptySource", "InvalidUrl", "FailedToLoadMedia"
   code: string | undefined
   stack?: string | undefined
}
```
that way the developer (myself) could use instanceof and throw

```javascript
if(error instanceof VideoError){
  // do something
}

throw new VideoError({
  name: "AwesomeError",
  message: "Thanks for this package, it works awesome"
})
```

**4. Add specific error typing for better management.**
The react-native-vision-camera library has added great type safety in their errors which really helps the developer handle different errors:
![image](https://github.com/react-native-video/react-native-video/assets/122328438/7fc77042-36be-454d-820e-36e7c0943a23)

### Alternatives considered

The author created their own local `VideoError` type as a workaround (shown in the Proposed solution above). They also cite react-native-vision-camera's typed error system as a reference approach.

### Platforms

iOS, Android

### Additional context

**Possible implementation:** Exactly as stated above.

**Code sample** — in summary, I would love to be able to handle errors exactly like so:

```javascript

onError((error: VideoError) => {
   switch(error.name) {
       case "EmptySource": // typescript intellisense here on the case statement
           alert("Your video source is empty")
           break;
       case "InvalidMedia":
           alert("The video you are trying to play is not supported")
           break;
       case "AwesomePackage":
           alert("The react native video package is sick af")
           break;
        default:
            alert(error.message);
    }

})
```

**Comment by KrzysztofMoch** [comment]
I have this on my mind - I will add this to roadmap to don't miss it!

**Comment by ChristopherGabba** [comment]
Thanks @KrzysztofMoch!

**Comment by TWG-AI** [comment]
v7 ships with a fully typed error system — `VideoError` is a generic class with a `VideoErrorCode` union (e.g. `'source/invalid-uri'`, `'player/not-initialized'`):

```ts
player.addEventListener('onError', (error: VideoRuntimeError) => {
  console.log(error.code);    // typed VideoErrorCode
  console.log(error.message); // human-readable string
});
```

> I'm an AI — check the v7 source for the full `VideoErrorCode` union.
